### DISCUSSION TITLE
Computed property vs Initialising

### DISCUSSION THREAD

**Discussion by Namedix** [pointfreeco/swift-dependencies#131] [Q&A] [answered]
Some time ago I noticed that in some situations we initialise clients  and in others we use them as computed properties, can someone explain to me when we should choose one solution over the other?

Computed property approach:
(Speech Client, SyncUps https://github.com/pointfreeco/syncups/blob/main/SyncUps-StackBased/SyncUps/Dependencies/SpeechClient.swift)
```swift
  static var liveValue: SpeechClient {
    let speech = Speech()
    return SpeechClient(
      authorizationStatus: { SFSpeechRecognizer.authorizationStatus() },
      requestAuthorization: {
        await withUnsafeContinuation { continuation in
          SFSpeechRecognizer.requestAuthorization { status in
            continuation.resume(returning: status)
          }
        }
      },
      startTask: { request in
        await speech.startTask(request: request)
      }
    )
  }
```

Initialising approach: 
(DataManager, SyncUps, https://github.com/pointfreeco/syncups/blob/main/SyncUps-StackBased/SyncUps/Dependencies/DataManager.swift)
```swift
  static let liveValue = DataManager(
    load: { url in try Data(contentsOf: url) },
    save: { data, url in try data.write(to: url) }
  )
```

So far, I created all my clients as computed properties, but I ran into problems where my client started to be initialized more than once. How should we do it?

**Comment by mbrandonw** [comment]
Hi @Namedix, dependencies should be cached the first time they are accessed. Can you provide a demonstration of a dependency being initialized multiple times?

**Reply by Namedix** [reply]
Thank you @mbrandonw for answer, I managed to find out why our libraries were initialized more than once, namely we use Tuist to modularize our project and Tuist Dependencies by default created static frameworks from SPM libraries, which is why our targets had more than one Dependencies implementation that initialized / override new values for us. 
![image](https://github.com/pointfreeco/swift-dependencies/assets/20950523/12e5613c-a4a9-4bd8-8f96-5ced2b930683)
Where when we initialized the client, it did not create it a second time. The change to dynamic frameworks helped us and now everything have only one implementation and works as expected! 


Although I would like to know when we should initialize values and when not, is there any rule we should follow?

**Reply by mbrandonw** [reply, marked as answer]
Hi @Namedix, good to hear, and we also discuss that gotcha in the [docs](https://pointfreeco.github.io/swift-dependencies/main/documentation/dependencies/testing/#Statically-linking-your-tests-target-to-Dependencies).

> Although I would like to know when we should initialize values and when not, is there any rule we should follow?

Both styles are fine to use and are legitimate. But sometimes only the computed static property style is possible because you need to capture some state on the inside, like in the `SpeechClient`. But if you don't need to capture that state, then a `static let` is just fine. Or if you prefer to coalesce on a single style for consistency, then feel free to use `static var` all the time.
